pkg/router/template/util/util.go

```diff
@@ -94,3 +104,130 @@ func GenerateBackendNamePrefix(termination routev1.TLSTerminationType) string {
 
 	return prefix
 }
+
+var haproxyDurationRE = regexp.MustCompile(`^([0-9]+)(us|ms|s|m|h|d)?$`)
+
+// HaproxyMaxTimeoutDuration is HaproxyMaxTimeout as a time.Duration value.
+var HaproxyMaxTimeoutDuration = func() time.Duration {
+	d, err := time.ParseDuration(HaproxyMaxTimeout)
+	if err != nil {
+		panic(err)
+	}
+	return d
+}()
+
+// OverflowError represents an overflow error from ParseHAProxyDuration.
+// OverflowError is returned if the value is greater than what time.ParseDuration
+// allows (value must be representable as uint64, e.g. approximately 2562047.79 hours
+// or 9223372036854775807 nanoseconds).
+type OverflowError struct {
+	error
+}
+
+// InvalidInputError represents an error based on invalid input to ParseHAProxyDuration
+type InvalidInputError struct {
+	error
+}
+
+// ParseHAProxyDuration is similar to time.ParseDuration, but modified with meaningful
+// error messages for invalid input, with support for decimal points removed, with
+// support for the "d" unit for days, and with "ms" as the default unit.
+// It parses a duration string and returns a suitable duration. A duration string is a
+// sequence of decimal numbers followed by a unit suffix,
+// such as "300ms", "1h" or "45m".  If the unit suffix is omitted, "ms" is used.
+// Valid time units are "us", "ms", "s", "m", "h", and "d".
+func ParseHAProxyDuration(s string) (time.Duration, error) {
+	orig := s
+	var d uint64
+
+	// Compare the input to the HAProxy duration regex.
+	if !haproxyDurationRE.MatchString(s) {
+		return 0, InvalidInputError{errors.New("invalid duration, no pattern match " + s)}
+	}
+
+	if s == "0" {
+		return 0, nil
+	}
+	for s != "" {
+		// The next character must be [0-9].
+		if !('0' <= s[0] && s[0] <= '9') {
+			return 0, InvalidInputError{errors.New("invalid duration, not a number " + orig)}
+		}
+
+		// Consume [0-9]*.
+		var v uint64
+		var err error
+		v, s, err = leadingInt(s)
+		if err != nil {
+			// overflow.
+			return 0, OverflowError{errors.New(err.Error() + orig)}
+		}
+
+		// Set the default HAProxy unit of "ms", in case a unit wasn't provided.
+		u := "ms"
+		// Consume unit.
+		i := 0
+		for ; i < len(s); i++ {
+			c := s[i]
+			if '0' <= c && c <= '9' {
+				break
+			}
+		}
+		if i > 0 {
+			u = s[:i]
+		}
+		s = s[i:]
+		unit, ok := unitMap[u]
+		if !ok {
+			return 0, InvalidInputError{errors.New("unknown unit " + u + " in duration " + orig)}
```

Comment:
This checks if the unit string is in the map, and you're right we don't need to verify that since the regexp does.   I'll remove the check to make sure the unit exists.